### ISSUE TITLE
Date field should have a style generator

### ISSUE THREAD

**Issue by pwilkin** [vaadin/framework#10856]
The current API for date field styling allows for styling specific dates, but it does not allow for styling recurring patterns (for example, having a separate style for Saturdays and Sundays). Using `setDateStyle(...)` is not really viable here since one would have to manually set styles for all Saturdays and Sundays for an arbitrary time period forward and backward. 

The solution would be to add a `setDateStyleGenerator(...)`, similar to what is currently preset on `Grid`.

**Comment by tsuoanttila** [comment]
We were discussing this approach when the `setDateStyle` API was introduced. The main issue was to figure out which dates to run the generator for. There's no callback to server when switching the calendar month, so generating those per month is not really feasible. Also if we run the generator for all possible dates between epoch and some arbitrary date, how to pick the arbitrary date?

There are some design questions that need answering before we can implement this.

As a workaround for now I'd suggest using the following approach:
```
DateField dateField = new DateField();
// This is the generator you want to use, return `null` for no style name
Function<LocalDate, String> styleGenerator = date -> date.getMonth().toString();
LocalDate end = LocalDate.of(2020, 01, 01); // Arbitrary start date
LocalDate current = LocalDate.of(2010, 01, 01); // Arbitrary end date
while (current.isBefore(end)) {
	String style = styleGenerator.apply(current);
	dateField.setDateStyle(current, style);
	current = current.plusDays(1);
}
```

**Comment by pwilkin** [comment]
Yes, but that is terribly inefficient (creating styles for 4000 days in advance?). 

A quick analysis of the client-side code shows that this could be implemented in the following manner instead:
* an RPC that calls the server-side requesting styles for the current month
* the generator on the server-side is always run on-demand for the requested month to avoid overhead

Of course, as you said, this would require a callback to the server when switching months, but that would come naturally since the client-side RPC needs to request the dates from a month anyways (if you need to add an RPC call, plugging it on the month change event isn't really a problem).

**Comment by elmot** [comment]
Looks like a good topic for an addon. Any volunteers? If something needs to be unlocked in the framework (refactored, converted private -> protected etc) just let us know.

**Comment by pwilkin** [comment]
Doesn't really look like addon material to me because you need to change the client code of `DateField` - don't see how an addon would go around doing that.

**Comment by elmot** [comment]
Addon can contain client-side code. The addon may be done by subclassing of existing classes (which may be pretty challenging in case of date fields), or more tricky way via `AbstractExtension`.
